tools/bundletool/bundletool.py

```diff
@@ -193,14 +193,13 @@ def _write_entry(self, dest, data, executable, out_zip, compress):
       BundleToolError: If two files with different content would be placed
           at the same location in the ZIP file.
     """
-    new_hash = hashlib.md5(data).digest()
-    existing_hash = self._entry_hashes.get(dest)
-    if existing_hash:
-      if existing_hash == new_hash:
+    existing_data = self._entry_datas.get(dest, None)
+    if existing_data is not None:
+      if hashlib.md5(existing_data).digest() == hashlib.md5(data).digest():
```

Comment:
drive by comments -

The hashes were all about reducing the amount of data in memory; once you are keeping the data just compare the data and never hash. The hash is just wasted work compared to a bytes compare.

Reply:
Yeah, the memory usage is the main downside I can think of. Another way to do this is to add an additional pass to iterate all the entries that need to be written first, and find if there are multiple files sharing the same dest. I can give it a try.

Reply:
what if we just stored the paths that have already been written, and lazily do `out_zip.read(path)` to get the data if the path already exists, would that work? also at that point if you were worried about it we could cache only the data from the conflicts in memory, but might not even be worth it

Reply:
I actually tried the additional pass solution. First, count which files have the same destination path, and then only calculate the hash for these files. I did a quick dirty implementation, and it worked. But I'm busy setting up buildfarm for our project, so I haven't got time to sort out the code and commit it. I will try to upload it within the next week.